### ISSUE TITLE
Job failed to load: undefined method `load_dj' for Syck:Module. 

### ISSUE THREAD

**Issue by manjunath-nm89** [collectiveidea/delayed_job#786]
Submitted a job and ran a DJ worker and the job failed with an exception `Job failed to load: undefined method 'load_dj' for Syck:Module. `

##### This is my configuration

- delayed_job version - 4.0.6
- delayed_job_active_record - 4.0.3
- Rails - 3.1.12
- YAML ENGINE: `YAML::ENGINE.yamler= 'syck'`

##### Exception Trace

Job failed to load: undefined method `load_dj' for Syck:Module. 
Handler: "--- !ruby/object:Delayed::PerformableMethod \nargs: \n- 198\n- 2015-02-25 01:08:41.226804 Z\nmethod_name: :set_favorite_last_accessed\nobject: !ruby/class SegmentReportQueriesController\n"

**Comment by albus522** [comment]
You have to set the YAML engine before DJ is required or the correct files will not be loaded. This means you have to set the YAML engine at the top of your `config/environment.rb` file before the `Bundler.require` line.

Also be aware many libraries are dropping Syck support and we probably will as well.

**Comment by manjunath-nm89** [comment]
I had the YAML ENGINE configuration in my `boot.rb` and tried with `environment.rb` as well. Both of them did not work. Digging into the gem now.

**Comment by manjunath-nm89** [comment]
I also checked that before the delayed job library is loaded, the YAML engine is correctly set to Syck

**Comment by manjunath-nm89** [comment]
The below code is from `delayed/backend/base.rb` For some reason the `def load_dj` defined in `syck_ext.rb` is not getting included.

``` ruby
def payload_object
  @payload_object ||= YAML.load_dj(handler)
  rescue TypeError, LoadError, NameError, ArgumentError, SyntaxError, Psych::SyntaxError => e
  raise DeserializationError, "Job failed to load: #{e.message}. Handler: #{handler.inspect}"
end
```

**Comment by albus522** [comment]
What version of ruby?

**Comment by manjunath-nm89** [comment]
`ruby 1.9.3p484`

**Comment by albus522** [comment]
So Syck detection does not work in 1.9.3. I can't test if 1.9.2 worked, and 2.0 removed Syck support. Given that 1.9.3 is now EOL and you are the only person with this issue in 3 years, our fix is probably going to be to remove the Syck extensions entirely, which doesn't help you, but If you fork and change https://github.com/collectiveidea/delayed_job/blob/master/lib/delayed/yaml_ext.rb#L5 to
```
if /syck|yecht/i.match(YAML.parser.class.name)
```
it should load the correct libraries but it still may not work.

**Comment by manjunath-nm89** [comment]
Yeah, I had already tried your fork and code change and it did not work. I have monkey patched a fix for now. We will be moving to `Ruby 2.0` shortly.

Thanks for your prompt responses.

**Comment by CarlosCD** [comment]
Hey Manjunath,

I am not sure if this is the fix you implemented, but this works well for me (added in a Rails initializer):

```ruby
module Delayed
  module Backend
    module Base
      def payload_object
        if YAML.respond_to? :load_dj
          @payload_object ||= YAML.load_dj(handler)
        elsif YAML.respond_to? :unsafe_load
          @payload_object ||= YAML.load(handler, safe: false)
        else
          @payload_object ||= YAML.load(handler)
        end
      rescue TypeError, LoadError, NameError, ArgumentError, SyntaxError, Psych::SyntaxError => e
        raise DeserializationError, "Job failed to load: #{e.message}. Handler: #{handler.inspect}"
      end
    end
  end
end
```

I am using:

* Ruby 1.9.3-p484 (hopefully migrating soon to 2.2.2, if I can get my way)
* Rails 4.2.3
* delayed_job_active_record 4.0.3 ([monkey patched for PostgreSQL 9.0](https://github.com/CarlosCD/delayed_job_active_record))
* delayed_job 4.0.6

**Comment by manjunath-nm89** [comment]
Carlos,

This is kind of the same fix which I monkey-patched and it works well for me. We have moved to Ruby 2.1 though and the monkey-patch is no longer needed.
